src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -340,10 +338,26 @@ void FileSegment::setRemoteFileReader(RemoteFileReaderPtr remote_file_reader_)
     remote_file_reader = remote_file_reader_;
 }
 
-void FileSegment::write(char * from, size_t size, size_t offset)
+void FileSegment::write(char * from, size_t size, size_t offset_in_file)
 {
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FileSegmentWriteMicroseconds);
+    assertWriteAllowed(size, offset_in_file);
+    writeImpl(size, offset_in_file, [&]()
+    {
+        if (!cache_writer)
+            cache_writer = std::make_unique<WriteBufferFromFile>(getPath(), /* buf_size */0);
 
+        /// Size is equal to offset as offset for write buffer points to data end.
+        cache_writer->set(from, /* size */size, /* offset */size);
+        /// Reset the buffer when finished.
+        SCOPE_EXIT({ cache_writer->set(nullptr, 0); });
+        /// Flush the buffer.
+        cache_writer->next();
```

Comment:
Can't we make `cache_writer` a local variable and re-create, since we reset it each time?

Reply:
There is still an opened fd in ctor of the write buffer, which may be useful to have already created as I remember there was sometimes contention in OpenedFileCache, and also there is a `sync` method in WriteBufferToFileSegment, which will be strangely implemented probably if we do not have a defined write buffer.